### DISCUSSION TITLE
`Annotated` doesn't work with global dependencies

### DISCUSSION THREAD

**Discussion by br3ndonland** [fastapi/fastapi#9674] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Annotated

from fastapi import Depends, FastAPI
from fastapi.testclient import TestClient


async def dependency() -> str:
    return "ok"


AnnotatedDependency = Annotated[str, Depends(dependency)]
app = FastAPI(dependencies=[AnnotatedDependency])


@app.get("/")
async def route_that_says_hello() -> str:
    return "Hello, World!"


def test_route_that_says_hello() -> None:
    test_client_app = TestClient(app)
    response = test_client_app.get("/")
    response_json = response.json()
    assert response_json == "Hello, World!"
```

### Description

Related: https://github.com/tiangolo/fastapi/discussions/9654

[FastAPI 0.95](https://github.com/tiangolo/fastapi/releases/tag/0.95.0) and https://github.com/tiangolo/fastapi/pull/4871 introduced support for `Annotated` dependencies, which are dependencies specified using `typing.Annotated` instead of as path operation function arguments. One of the benefits is that `Annotated` dependencies can be re-used more easily.

Another feature that FastAPI provides for re-using dependencies is [global dependencies](https://fastapi.tiangolo.com/tutorial/dependencies/global-dependencies/). These are dependencies that apply to all path operation functions of the `FastAPI` or `APIRouter` instance. Global dependencies were introduced in [FastAPI 0.62](https://github.com/tiangolo/fastapi/releases/tag/0.62.0).

`Annotated` dependencies seem to work well when they are used in individual path operation functions, but they don't seem to work as global dependencies.

To reproduce this, the provided example sets up a simple test app with a global dependency, a path operation function that uses the dependency, and a test for the path operation function.

When attempting to run the test, the following error will be seen:

```text
test_app.py:15: in <module>
    @app.get("/")
fastapi/routing.py:669: in decorator
    self.add_api_route(
fastapi/routing.py:608: in add_api_route
    route = route_class(
fastapi/routing.py:458: in __init__
    get_parameterless_sub_dependant(depends=depends, path=self.path_format),
fastapi/dependencies/utils.py:134: in get_parameterless_sub_dependant
    depends.dependency
/opt/homebrew/Cellar/python@3.11/3.11.3/Frameworks/Python.framework/Versions/3.11/lib/python3.11/typing.py:2138: in __getattr__
    return super().__getattr__(attr)
/opt/homebrew/Cellar/python@3.11/3.11.3/Frameworks/Python.framework/Versions/3.11/lib/python3.11/typing.py:1281: in __getattr__
    return getattr(self.__origin__, attr)
E   AttributeError: type object 'str' has no attribute 'dependency'
```

Simply changing the app to `app = FastAPI(dependencies=[Depends(dependency)])` resolves the error.

Based on how FastAPI injects dependencies, it appears that we can't add `Annotated` top-level dependencies:

- When instantiating `FastAPI` like `FastAPI(dependencies=[AnnotatedDependency])`
- When instantiating `APIRouter` like `APIRouter(dependencies=[AnnotatedDependency])`
- When including `APIRouter` instances in a `FastAPI` instance like `app_instance.include_router(router_instance, dependencies=[AnnotatedDependency])`

The docs are unclear about this. PR https://github.com/tiangolo/fastapi/pull/9268 claimed to "Update all docs to use `Annotated` as the main recommendation," but it did not. This is particularly confusing in the docs on "[Bigger Applications - Multiple Files](https://fastapi.tiangolo.com/tutorial/bigger-applications/)" (where `APIRouter` is introduced). There's a short "Dependencies" section that uses `Annotated`, but then immediately after that we see the old-style dependency specification when instantiating `FastAPI` and `APIRouter`.

https://github.com/tiangolo/fastapi/blob/32935103b12b1548117abef0b4af9dd883898308/docs_src/bigger_applications/app_an/main.py#L7-L18

Am I missing some context here? Am I making a mistake with how I use `Annotated`?

If I've identified a limitation of FastAPI, I'm happy to contribute PRs to fix the code and docs.

### Operating System

Linux, macOS

### FastAPI Version

0.97.0

### Python Version

3.11.3

**Comment by ahmedabdou14** [comment]
Hi, this issue has been reported in this discussion https://github.com/tiangolo/fastapi/discussions/9654 too.

**Reply by br3ndonland** [reply]
> Hi, this issue has been reported in this discussion #9654 too.

I linked to that discussion in the description. Similar but not the same. #9654 is about "[dependencies in path operation decorators](https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-in-path-operation-decorators/)," this discussion is about "[global dependencies](https://fastapi.tiangolo.com/tutorial/dependencies/global-dependencies/)." They are on two different pages of the docs but may use the same dependency injection utils under the hood.

**Comment by Kludex** [comment, marked as answer]
An annotated type is a type, not a `Depends` object. 

In other words: `AnnotatedDependency = Annotated[str, Depends(dependency)]` is a `str` with a metadata `Depends(dependency)`.

**Reply by ahmedabdou14** [reply]
Actually this is a very important feature to be added to FastApi to provide consistency when using Depends. Since now it is recommended to use annotated, it might be very useful to add support for it in global dependencies and route dependencies aswell.

**Reply by Kludex** [reply]
Those are different things:

```py
from typing import Annotated

from fastapi import Depends, FastAPI

app = FastAPI()


async def common_parameters(q: str | None = None, skip: int = 0, limit: int = 100):
    return {"q": q, "skip": skip, "limit": limit}


CommonsDep = Annotated[dict, Depends(common_parameters)]


@app.get("/items/")
async def read_items(commons: CommonsDep):  # Here commons is being type hinted as CommonsDep, and there's no default value.
    return commons
```
